runtime/tr.source/trj9/env/VMJ9.cpp

```diff
@@ -1335,6 +1335,81 @@ TR_J9VMBase::compareAndSwapInt64FieldAt(uintptrj_t objectPointer, uintptrj_t fie
    return success != 0;
    }
 
+uint32_t
+TR_J9VMBase::getArrayElementSize(TR::DataType type)
+   {
+   if (type == TR::Address)
+      return TR::comp()->useCompressedPointers() ? sizeof(uint32_t) : sizeof(uintptrj_t);
+   else
+      return TR::Symbol::convertTypeToSize(type);
+   }
+
+bool
+TR_J9VMBase::isDiscontiguousArray(uintptrj_t objectPointer)
+   {
+   TR_ASSERT(haveAccess(), "isDicontiguousArray requires VM access");
+   TR_ASSERT(isClassArray(getObjectClass(objectPointer)), "Object is not an array");
+
+   int32_t length = *(int32_t*)(objectPointer + getOffsetOfContiguousArraySizeField());
+
+   if (TR::Compiler->om.canGenerateArraylets()
+       && length == 0)
+      return true;
+
+   return false;
+   }
+
+uintptrj_t
+TR_J9VMBase::getAddressOfElementFromContiguousArray(uintptrj_t objectPointer, TR::DataType type, int32_t elementIndex)
+   {
+   TR_ASSERT(haveAccess(), "getDataStartOfArray requires VM access");
+   TR_ASSERT(!isDiscontiguousArray(objectPointer), "Array is not contiguous");
+   uint32_t elementSize = getArrayElementSize(type);
+   return objectPointer + TR::Compiler->om.contiguousArrayHeaderSizeInBytes() + elementSize * elementIndex;
+   }
+
+uintptrj_t
+TR_J9VMBase::getAddressOfElementFromDiscontiguousArray(uintptrj_t objectPointer, TR::DataType type, int32_t elementIndex)
+   {
+   TR_ASSERT(haveAccess(), "getDataStartOfDiscontiguousArray requires VM access");
+   TR_ASSERT(TR::Compiler->om.canGenerateArraylets() && isDiscontiguousArray(objectPointer), "Array is not discontiguous");
+
+   uint32_t elementSize = getArrayElementSize(type);
+   uint32_t leafIndex = getArrayletLeafIndex(elementIndex, elementSize);
+   uint32_t elementIndexInLeaf = getLeafElementIndex(elementIndex, elementSize);
+   uintptrj_t dataStart = objectPointer + TR::Compiler->om.discontiguousArrayHeaderSizeInBytes();
+
+   if (TR::comp()->useCompressedPointers())
+      {
+      uint32_t *spine = (uint32_t*)dataStart;
+      dataStart = spine[leafIndex];
+      dataStart = decompressReference(dataStart);
+      }
+   else
+      {
+      uintptrj_t *spine = (uintptrj_t*)dataStart;
+      dataStart = spine[leafIndex];
+      }
+
+   return dataStart + elementIndexInLeaf * elementSize;
```

Comment:
I suspect there's a good upper bound on `om.arrayletLeafLogSize()`, in which case this multiplication should be safe as is